File: cloud-reactor/src/test/java/com/sequenceiq/cloudbreak/cloud/handler/AbstractComponentTest.java
package com.sequenceiq.cloudbreak.cloud.handler;

import jakarta.inject.Inject;

import org.junit.runner.RunWith;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;

import com.sequenceiq.cloudbreak.cloud.event.CloudPlatformRequest;
import com.sequenceiq.cloudbreak.cloud.handler.testcontext.TestApplicationContext;
import com.sequenceiq.cloudbreak.eventbus.Event;
import com.sequenceiq.cloudbreak.eventbus.EventBus;

@RunWith(SpringJUnit4ClassRunner.class)
@SpringBootTest(classes = TestApplicationContext.class,
        properties = "spring.main.allow-bean-definition-overriding=true")
public abstract class AbstractComponentTest<T> {

    @Inject
    private EventBus eb;

    @Inject
    private ParameterGenerator g;

    protected T sendCloudRequest() {
        CloudPlatformRequest<T> request = getRequest();
        return sendCloudRequest(request);
    }

    protected T sendCloudRequest(CloudPlatformRequest<T> request) {
        eb.notify(getTopicName(), Event.wrap(request));
        try {
            return request.await();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new RuntimeException(e);
        }
    }

    protected abstract String getTopicName();

    protected abstract CloudPlatformRequest<T> getRequest();

    protected ParameterGenerator g() {
        return g;
    }

}

File: cloud-reactor/src/test/java/com/sequenceiq/cloudbreak/cloud/handler/AbstractComponentTest.kt
package com.sequenceiq.cloudbreak.cloud.handler

import javax.inject.Inject

import org.junit.runner.RunWith
import org.springframework.boot.test.SpringApplicationConfiguration
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner

import com.sequenceiq.cloudbreak.cloud.event.CloudPlatformRequest
import com.sequenceiq.cloudbreak.cloud.handler.testcontext.TestApplicationContext

import reactor.bus.Event
import reactor.bus.EventBus

@RunWith(SpringJUnit4ClassRunner::class)
@SpringApplicationConfiguration(classes = TestApplicationContext::class)
abstract class AbstractComponentTest<T> {
    @Inject
    private val eb: EventBus? = null

    @Inject
    private val g: ParameterGenerator? = null

    protected fun sendCloudRequest(): T {
        val request = request
        return sendCloudRequest(request)
    }

    protected fun sendCloudRequest(request: CloudPlatformRequest<Any>): T {
        eb!!.notify(topicName, Event.wrap(request))
        try {
            return request.await() as T
        } catch (e: InterruptedException) {
            throw RuntimeException(e)
        }

    }

    protected abstract val topicName: String

    protected abstract val request: CloudPlatformRequest<Any>

    protected fun g(): ParameterGenerator {
        return g
    }

}
